main.go

```diff
@@ -0,0 +1,239 @@
+		return
+	}
+
+	c.IndentedJSON(http.StatusOK, message)
+}
+
+func createMessage(c *gin.Context) {
+	var newMsg message
+
+	if err := c.BindJSON(&newMsg); err != nil {
+		return
+	}
+
+	newMsg.UUID = uuid.NewString()
+	newMsg.Created = time.Now().Format(time.RFC3339Nano)
+	newMsg.Type = "message"
+	newMsg.Upvote = 0
+	newMsg.Downvote = 0
+	newMsg.CommentCount = 0
+	newMsg.Deleted = false
+
+	messages = append(messages, newMsg)
+	c.IndentedJSON(http.StatusCreated, newMsg)
+}
+
+func addMessageUpvote(c *gin.Context) {
+	id, ok := c.GetQuery("id")
+
+	if !ok {
+		c.IndentedJSON(http.StatusBadRequest, gin.H{"message": "Missing id query parameter"})
+	}
+
+	message, err := helperGetMessageById(id)
+
+	if err != nil {
+		c.IndentedJSON(http.StatusNotFound, gin.H{"message": "Message not found!"})
+		return
+	}
+
+	message.Upvote += 1
+	c.IndentedJSON(http.StatusOK, message)
+}
+
+func removeMessageUpvote(c *gin.Context) {
+	id, ok := c.GetQuery("id")
+
+	if !ok {
+		c.IndentedJSON(http.StatusBadRequest, gin.H{"message": "Missing id query parameter"})
+	}
+
+	message, err := helperGetMessageById(id)
+
+	if err != nil {
+		c.IndentedJSON(http.StatusNotFound, gin.H{"message": "Message not found!"})
+		return
+	}
+
+	if message.Downvote <= 0 {
+		c.IndentedJSON(http.StatusBadRequest, gin.H{"message": "No upvotes to remove!"})
+		return
+	}
+
+	message.Downvote -= 1
+	c.IndentedJSON(http.StatusOK, message)
+}
+
+func addMessageDownvote(c *gin.Context) {
+	id, ok := c.GetQuery("id")
+
+	if !ok {
+		c.IndentedJSON(http.StatusBadRequest, gin.H{"message": "Missing id query parameter"})
+	}
+
+	message, err := helperGetMessageById(id)
+
+	if err != nil {
+		c.IndentedJSON(http.StatusNotFound, gin.H{"message": "Message not found!"})
+		return
+	}
+
+	message.Downvote += 1
+	c.IndentedJSON(http.StatusOK, message)
+}
+
+func removeMessageDownvote(c *gin.Context) {
+	id, ok := c.GetQuery("id")
+
+	if !ok {
+		c.IndentedJSON(http.StatusBadRequest, gin.H{"message": "Missing id query parameter"})
+	}
+
+	message, err := helperGetMessageById(id)
+
+	if err != nil {
+		c.IndentedJSON(http.StatusNotFound, gin.H{"message": "Message not found!"})
+		return
+	}
+
+	if message.Downvote <= 0 {
+		c.IndentedJSON(http.StatusBadRequest, gin.H{"message": "No downvotes to remove!"})
+		return
+	}
+
+	message.Downvote -= 1
+	c.IndentedJSON(http.StatusOK, message)
+}
+
+func deleteMessage(c *gin.Context) {
+	id, ok := c.GetQuery("id")
+
+	if !ok {
+		c.IndentedJSON(http.StatusBadRequest, gin.H{"message": "Missing id query parameter"})
+	}
+
+	message, err := helperGetMessageById(id)
+
+	if err != nil {
+		c.IndentedJSON(http.StatusNotFound, gin.H{"message": "Message not found!"})
+		return
+	}
+
+	if message.Deleted {
+		c.IndentedJSON(http.StatusBadRequest, gin.H{"message": "Message has already beem deleted!"})
+		return
+	}
+
+	message.Deleted = true
+	c.IndentedJSON(http.StatusOK, message)
+}
+
+func updateMessage(c *gin.Context) {
```

Comment:
We only allow for updated Category, Title, and Message. Message update maintains the current message and appends 'Edit: ' plus the new message to the content.